modules/renter/repair.go

```diff
@@ -185,6 +202,39 @@ func (f *file) expiringContracts(height types.BlockHeight) []fileContract {
 	return expiring
 }
 
+// isOffline decides whether a host should be considered offline, based on its
+// scan metrics.
+func isOffline(host modules.HostDBEntry) bool {
+	// consider a host offline if:
+	// 1) The host has been scanned at least three times, and
+	// 2) The three most recent scans have all failed, and
+	// 3) The time between the most recent scan and the last successful scan
+	//    (or first scan) is at least uptimeWindow
+	numScans := len(host.ScanHistory)
+	if numScans < uptimeMinScans {
+		// not enough data to make a fair judgment
+		return false
+	}
+	// NOTE: ScanHistory is ordered from oldest-newest
+	recent := host.ScanHistory[numScans-uptimeMinScans:]
+	for _, scan := range recent {
+		if scan.Success {
+			// one of the scans succeeded
+			return false
+		}
+	}
+	// initialize window bounds
+	windowStart, windowEnd := host.ScanHistory[0].Timestamp, host.ScanHistory[numScans-1].Timestamp
+	// iterate from newest-oldest, seeking to last successful scan
+	for i := numScans - 1; i >= 0; i-- {
+		if scan := host.ScanHistory[i]; scan.Success {
+			windowStart = scan.Timestamp
+			break
+		}
+	}
+	return windowEnd.Sub(windowStart) >= uptimeWindow
+}
```

Comment:
A method of the contractor, you mean? Should it still be called `IsOffline`? In its current incarnation, `isOffline` only needs the `modules.HostDBEntry`, so it seems strange to make it part of the contractor. I can see your reasoning though if we expand it to a more general ruling about whether the contract is useful at all.